### DISCUSSION TITLE
Help modify diff behavior when using neovim as git difftool

### DISCUSSION THREAD

**Discussion by vlad-tim** [neovim/neovim#27512] [Q&A] [answered]
When I make a diff of two files using `nvimdiff` as a git difftool it does not look they way I want. More specifically, removed lines are visualized as added instead of being indicated as removed. When I run `git difftool main.md` this is what I see:

![ksnip_20240217-170625](https://github.com/neovim/neovim/assets/11474041/93e8bdb9-941d-426d-9592-b6dd73825700)

If I execute the same command using VSCode as difftool (i.e. `code --wait --diff $LOCAL $REMOTE), the result matches my expectations and looks like this:

![ksnip_20240217-170708](https://github.com/neovim/neovim/assets/11474041/ae3eba7f-8b4d-4287-b5df-18f65fd1e270)

Is there a way to achieve the desired look? I would like make removed lines have red background and added lines have green background.

**Comment by clason** [comment]
Just fix your colorscheme. (Use `:Inspect` to find out the used highlight groups you need to customize.)

**Comment by vlad-tim** [comment]
I tried to run `:Inspect` file keeping the cursor on the line with a problem, but it displays the error:

```
No items found at position 3,0 in buffer 1
```

I even tried to upgrade nvim to the nightly version, but result is the same.

**Reply by clason** [reply]
`:h hl-DiffDelete`

**Reply by vlad-tim** [reply]
I added `highlight DiffDelete guifg=None guibg=#ff0000` to the configuration, and result is the following:

![ksnip_20240217-174419](https://github.com/neovim/neovim/assets/11474041/2aefc165-e1ac-44e1-a2ac-63e68fca0a64)

The background color on the right side became brighter. With that said, it may be not a problem with the color scheme, but something else. May my help request be re-opened?

**Reply by clason** [reply]
What else could it be a problem with? It's unclear what exactly you're asking for here. The result you stated in your question can be achieved by highlighting `DiffDelete` and `DiffAdd`.

**Reply by vlad-tim** [reply]
Assuming @zeertzjq is right and both removed and added lines are `DiffAdd` , then it is not clear to me how to distinguish them and paint differently.

**Reply by clason** [reply]
Of course not, that would make zero sense. The added line is `DiffAdd`, and the removed line is `DiffDelete`.

**Comment by zeertzjq** [comment, marked as answer]
```
|hl-DiffAdd|	DiffAdd		Added (inserted) lines.  These lines exist in
				this buffer but not in another.
|hl-DiffDelete|	DiffDelete	Deleted lines.  Also called filler lines,
				because they don't really exist in this
				buffer.
```
These are not removed lines. They are added lines, so you should see `:h hl-DiffAdd` and `:h 'winhighlight'`. Try this:
```vim
autocmd BufWinEnter * if &diff && argidx() == 0 | setlocal winhighlight=DiffAdd:DiffDelete,DiffDelete:DiffAdd | endif
```

**Reply by vlad-tim** [reply]
My original screenshots showed how it looks when `DiffDelete` is painted red, and `DiffAdd` is painted green. I tried to make both `DiffDelete` and `DiffAdd` red and it led to the following.

![ksnip_20240217-175400](https://github.com/neovim/neovim/assets/11474041/5057e817-5ac9-4e06-9b5d-5b108f433a09)

And thus I believe that @zeertzjq is technically right. But the same diff looks correct in VSCode (and other IDEs, like IntelliJ)

![ksnip_20240217-175501](https://github.com/neovim/neovim/assets/11474041/2c93d433-e1f4-4ffc-9715-2e5840c279ed)

Is it possible to achieve this, and if yes, how?

**Reply by clason** [reply]
I still don't understand what you want. What _specifically_ from the second screenshots do you want to reproduce? The colors can be customized freely. (Of course, you don't want to make them _both_ red.)

**Reply by vlad-tim** [reply]
If looking at the last screenshot I posted (from VSCode), I specifically want:

1. make the line number 4 on the left side of the diff have red background in neovim
2. make the line number 5 on the right side of the diff have green background in neovim

**Reply by clason** [reply]
*Sigh.* This is now a different example than in your original question.

**Reply by vlad-tim** [reply]
I hope it helped  to clarify the question and avoid confusion. It is still unclear to me if there is a solution.

**Reply by clason** [reply]
I suspect not. You can customize the colors, but not which lines are shown.

**Reply by zeertzjq** [reply]
Try this:
```vim
autocmd BufWinEnter * if &diff && argidx() == 0 | setlocal winhighlight=DiffAdd:DiffDelete,DiffDelete:DiffAdd | endif
```

**Reply by vlad-tim** [reply]
Oh, wow, it did the trick!

![ksnip_20240217-182903](https://github.com/neovim/neovim/assets/11474041/e76554c1-5761-4977-9560-e919b8788201)

(I have been trying to wrap my head around `winhighlight` after reading you earlier message, but couldn't figure it out)

**Comment by bew** [comment]
I think the question is how diff highlights are computed: comparaison vs before/after

The diff in (n)vim shows you the difference compared to the other file. So the left file has 1 added line (in green) "compared to the right file", and the right file has 1 removed line "compared to the left file". But (n)vim doesn't have a concept of the 'current' file.

In git / vscode /.. it has this distinction of the 'current' file (right) and what it was before (left).
When showing the old file only deleted lines are highlighted, and on the 'current' file only new lines are highlighted.

**Reply by clason** [reply]
So you're not asking about the colors, but about the hatchings. For the deleted lines, see `:h 'fillchars'`. For the added lines, this is not possible.

**Reply by bew** [reply]
It's not about the hatching either, but about where deleted/added lines are shown.

There is no way currently afaik to only show deleted lines on the left (but not added lines) and added lines on the right (but not deleted lines).

In comparison vscode/git/delta.. would never show added lines on the left pane (text with green bg) or deleted lines on the right pane (text with red bg).

**Reply by clason** [reply]
Then, no, that is not possible.
